### PR TITLE
[WIP]Removing an additional job_save_db call for subjob's run path

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Describe Bug or Feature

When running a new subjob, we call job_save_db twice: once from inside req_runjob2->svr_startjob->..->svr_setjobstate, and once inside req_runjob2:
before

This is redundant, so we'll save some time if we only saved once if possible.

#### Describe Your Change

Moved the job_save_db call to after svr_setjobstate(). This causes svr_setjobstate() to skip calling job_save_db as it's a new subjob so that newobj bit is still set. So, we only save the subjob once.

#### Attach Test and Valgrind Logs/Output

Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|6471|3990|0|0|0|0|3990|

### CODE REVIEW

File: src/server/req_runjob.c [hunk 1]
```diff
@@ -771,7 +770,11 @@ req_runjob2(struct batch_request *preq, job *pjob)
 		/* provisioning was needed and was qneueued successfully */
 		/* Allways send ack for prov jobs, even if not async run */
 		reply_ack(preq);
-		return;
+		if (save_subjob(pjob)) {
+			free_nodes(pjob);
+			req_reject(PBSE_SAVE_ERR, 0, preq);
```
Comment (subhasisb) [inline src/server/req_runjob.c:775]: a req_reject() after a reply_ack() is not meaningful. I suggest we move the reply_ack() to after this block

File: src/server/req_runjob.c [hunk 2]
```diff
@@ -748,19 +758,8 @@ req_runjob2(struct batch_request *preq, job *pjob)
...
-	if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) {
-		if (job_save_db(pjob)) {
-			free_nodes(pjob);
-			req_reject(PBSE_SAVE_ERR, 0, preq);
-			return;
-		}
-	}
-
 	if (prov_rc) { /* problem with the request */
+		save_subjob(pjob);
```
Comment (subhasisb) [inline src/server/req_runjob.c:762]: if you fail to save, you want to reject with the DB save error?

File: src/server/req_runjob.c [hunk 3]
```diff
@@ -802,14 +807,20 @@ req_runjob2(struct batch_request *preq, job *pjob)
...
 		((rq_type == PBS_BATCH_AsyrunJob_ack) || preq)) {
+		save_subjob(pjob);
 		free_nodes(pjob);
-		if (preq)
-			req_reject(rc, 0, preq);
+		req_reject(rc, 0, preq);
+	}
+
+	/* if a subjob, the job wouldn't have gotten saved yet, let's save it now */
+	if (save_subjob(pjob)) {
```
Comment (subhasisb) [inline src/server/req_runjob.c:821]: u seem to have just shifted the location of the subjob save, how is it cutting a job save. Certainly the save inside of svr_setjobstate() is also still happening?

**Review by subhasisb — CHANGES_REQUESTED** [review]

File: src/server/req_runjob.c (hunk 1 above)
Comment (bhroam) [reply src/server/req_runjob.c:775]: Don't all of the reply/reject functions free the batch reply structure?  Won't doing 2 of them double free?

File: src/server/req_runjob.c (hunk 3 above)
Comment (agrawalravi90) [reply src/server/req_runjob.c:821]: svr_setjobstate() doesn't save to db if newobj is set, which it is for the subjob. Calling job_save_db() before svr_setjobstate() clears that bit and causes svr_setjobstate() to call job_sabe_db again. That's how moving the job_save_db call after svr_setjobstate keeps the bit set and svr_setjobstate doesn't call job_save_db. I've added this explanation to the PR description now.

File: src/server/req_runjob.c (hunk 2 above)
Comment (agrawalravi90) [reply src/server/req_runjob.c:762]: this was intentional, we are in an error case anyways, I think it's better to return the original error than return the db save error, that's why i didn't handle it. Let me know what you think.

File: src/server/req_runjob.c (hunk 3 above)
Comment (subhasisb) [reply src/server/req_runjob.c:820]: Thanks for explaining, makes sense.

File: src/server/req_runjob.c (hunk 2 above)
Comment (subhasisb) [reply src/server/req_runjob.c:762]: Sounds fine

File: src/server/req_runjob.c (hunk 3 above)
Comment (subhasisb) [reply src/server/req_runjob.c:820]: Could you add a comment explaining why we are not saving the subjob before and how setjobstate() will not save it - somebody else will certainly forget this detail (i myself added the check for newjobstate in setjobstate() but did not recall that)

**Review by subhasisb — CHANGES_REQUESTED** [review]

**Review by subhasisb — APPROVED** [review]

**Review by ashwathraop — APPROVED** [review]

**Review by subhasisb — APPROVED** [review]

**Comment by subhasisb** [comment]
@agrawalravi90 if the server process was to crash right after this state change to 41, can you describe what would happen in that case? In other words, we need to test for that case (gdb would be handy here)

File: src/server/svr_jobfunc.c [hunk 4]
```diff
@@ -506,7 +506,7 @@ svr_dequejob(job *pjob)
  */
 
 int
-svr_setjobstate(job *pjob, char newstate, int newsubstate)
+svr_setjobstate(job *pjob, char newstate, int newsubstate, bool db_write)
```
Comment (ashwathraop) [inline src/server/svr_jobfunc.c:509]: Update Doxygen with the new argument.

**Comment by agrawalravi90** [comment]
I have come across the following issue with this enhancement, for subjobs:
- if server process is killed before it's had a chance to save it to db (because of my change, the save to db happens after sending the job to mom), then the subjob is never saved to db
- So, when server wakes up, mom tells it about the subjob, but server doesn't have it (subjobs are created only when they are run), so it tells the mom to discard the job, and it ends up getting terminated

As far as I can tell, this is not trivial to fix. We'll need to recreate the subjob object and set all of the metadata on it that is set when a job is run. This means walking down some of the runjob path, but skipping the rest of it (since the array parent and nodes are already allocated). I'm still trying to figure out if this is worth the effort, in the meantime, if you guys have any thoughts/suggestions, do let me know. I'll investigate this for a week more at max and then decide whether to close this PR or not.

**Comment by agrawalravi90** [comment]
Another thought: instead of this, one, possibly lower hanging fruit, might be to not send exec_vnode for jobs from moms to server when doing registermom. The server doesn't use it, so we could save some data transfer and not send this over. Of course this would mean that there's no way to do the fix needed for this PR as we'll no longer have an exec_vnode from the mom, but since that's a messy route anyways, maybe it's better to go this other route.

**Pull request closed without merging**